Grid.cpp

```diff
@@ -168,103 +130,105 @@ Grid::Grid(MPI_Comm comm, const std::string& filename, const std::string& dim0_n
                 index++;
             }
         }
+    }
+
+    NC_CHECK(nc_close(nc_id));
+}
+
+Grid::Grid(MPI_Comm comm, const std::string& filename, const std::string& dim0_name,
+    const std::string& dim1_name, const std::string& mask_name, bool ignore_mask)
+    : _comm(comm)
+    , _dim0_name(dim0_name)
+    , _dim1_name(dim1_name)
+{
+
+    CHECK_MPI(MPI_Comm_rank(comm, &_rank));
+
+    Grid::ReadGridDims(filename);
+
+    // Initially we partition assuming there is no land mask
+    // Figure out my subset of objects
+    CHECK_MPI(MPI_Comm_size(comm, &_num_procs));
+
+    // Start from a 2D decomposition
+    _num_procs_0 = _num_procs;
+    _num_procs_1 = 1;
+    find_factors(_num_procs, _num_procs_0, _num_procs_1);
+
+    _local_ext_0 = ceil((float)_global_ext_0 / (_num_procs_0));
+    _local_ext_1 = ceil((float)_global_ext_1 / (_num_procs_1));
+    _global_0 = (_rank / _num_procs_1) * _local_ext_0;
+    _global_1 = (_rank % _num_procs_1) * _local_ext_1;
+
+    if ((_rank / _num_procs_1) == _num_procs_0 - 1) {
+        _local_ext_0 = _global_ext_0 - (_rank / _num_procs_1) * _local_ext_0;
+    }
+    if ((_rank % _num_procs_1) == _num_procs_1 - 1) {
+        _local_ext_1 = _global_ext_1 - (_rank % _num_procs_1) * _local_ext_1;
+    }
```

Comment:
Since you define `_global_0` and `_global_1`, you could use them inside the `if` statements rather than recomputing.